File: frontend/src/services/applications.ts
import { API_BASE_URL, ApiError, apiRequest } from "./api";

export type ApplicationListItem = {
  id: string;
  job_id: string;
  title: string;
  company_name: string;
  status: string;
  match_score: number;
  applied_at: string | null;
  latest_email_classification: string | null;
};

export type EmailMonitorData = {
  gmail_thread_id: string;
  sender: string;
  subject: string;
  snippet: string;
  latest_classification: string;
  last_checked_at: string;
  is_resolved: boolean;
};

export type ApplicationDetail = {
  id: string;
  job_id: string;
  title: string;
  company_name: string;
  status: string;
  match_score: number;
  cover_letter_text: string;
  ats_provider: string | null;
  confirmation_url: string | null;
  screenshot_urls: string[];
  email_monitor: EmailMonitorData | null;
};

export type ApplicationsListResponse = {
  items: ApplicationListItem[];
};

export type SourcePerformanceItem = {
  source: string;
  total_applications: number;
  replied_count: number;
  response_rate: number;
};

export type TitlePerformanceItem = {
  title: string;
  total_applications: number;
  replied_count: number;
  response_rate: number;
};

export type ApplicationsStats = {
  total_applications: number;
  total_applied: number;
  total_replied: number;
  response_rate: number;
  avg_hours_to_first_reply: number | null;
  source_breakdown: SourcePerformanceItem[];
  top_titles: TitlePerformanceItem[];
};

export type NotificationItem = {
  application_id: string;
  company_name: string;
  title: string;
  classification: string;
  snippet: string;
  created_at: string;
};

export type NotificationsResponse = {
  items: NotificationItem[];
};

export async function getApplications(): Promise<ApplicationsListResponse> {
  return apiRequest<ApplicationsListResponse>("/api/v1/applications");
}

export async function getApplicationDetail(applicationId: string): Promise<ApplicationDetail> {
  return apiRequest<ApplicationDetail>(`/api/v1/applications/${applicationId}`);
}

export async function getApplicationsStats(): Promise<ApplicationsStats> {
  return apiRequest<ApplicationsStats>("/api/v1/applications/stats");
}

export async function getNotifications(): Promise<NotificationsResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/notifications`, {
    credentials: "include",
  });

  if (!response.ok) {
    throw new ApiError("Unable to load notifications", response.status);
  }

  const payload = await response.text();
  const dataLine = payload
    .split("\n")
    .map((line) => line.trim())
    .find((line) => line.startsWith("data:"));

  if (!dataLine) {
    throw new ApiError("Notifications stream returned no data", response.status);
  }

  return JSON.parse(dataLine.slice(5).trim()) as NotificationsResponse;
}

export function subscribeNotifications(
  onNotifications: (notifications: NotificationsResponse) => void,
  onError?: (error: Error) => void,
): () => void {
  const source = new EventSource(`${API_BASE_URL}/api/v1/notifications/stream`, {
    withCredentials: true,
  });
  let closed = false;

  const handleNotifications = (event: MessageEvent<string>) => {
    const notifications = JSON.parse(event.data) as NotificationsResponse;
    onNotifications(notifications);
  };

  source.addEventListener("notifications", handleNotifications as EventListener);
  source.onerror = () => {
    if (closed) {
      return;
    }
    if (source.readyState === EventSource.CLOSED) {
      onError?.(new ApiError("Notifications stream disconnected", 503));
    }
  };

  return () => {
    closed = true;
    source.close();
  };
}

File: backend/app/schemas/applications.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class EmailMonitorData(BaseModel):
    gmail_thread_id: str
    sender: str
    subject: str
    snippet: str
    latest_classification: str
    last_checked_at: datetime
    is_resolved: bool


class ApplicationListItem(BaseModel):
    id: str
    job_id: str
    title: str
    company_name: str
    status: str
    match_score: float
    applied_at: datetime | None
    is_demo: bool = False
    latest_email_classification: str | None = None


class ApplicationsListData(BaseModel):
    items: list[ApplicationListItem] = Field(default_factory=list)


class SourcePerformanceItem(BaseModel):
    source: str
    total_applications: int
    replied_count: int
    response_rate: float


class TitlePerformanceItem(BaseModel):
    title: str
    total_applications: int
    replied_count: int
    response_rate: float


class ApplicationsStatsData(BaseModel):
    total_applications: int
    total_applied: int
    total_replied: int
    response_rate: float
    avg_hours_to_first_reply: float | None = None
    source_breakdown: list[SourcePerformanceItem] = Field(default_factory=list)
    top_titles: list[TitlePerformanceItem] = Field(default_factory=list)


class ApplicationDetailData(BaseModel):
    id: str
    job_id: str
    title: str
    company_name: str
    status: str
    match_score: float
    cover_letter_text: str
    ats_provider: str | None = None
    confirmation_url: str | None = None
    confirmation_number: str | None = None
    is_demo: bool = False
    screenshot_urls: list[str] = Field(default_factory=list)
    email_monitor: EmailMonitorData | None = None


class ApplicationStatusUpdatePayload(BaseModel):
    status: Literal["interview_requested", "rejected", "offer", "withdrawn"]


class ApplicationStatusUpdateData(BaseModel):
    application_id: str
    status: str
